browser/scripts/ui-core.js

```diff
@@ -98,131 +83,201 @@ VizorUI.prototype._init = function(e2) {	// called by .init() in ui.js
-			var key;
-			for (key in domPanelStates) {
-				if (domPanelStates.hasOwnProperty(key)) {
-					ps[key] = VizorUI.getDomPanelState(domPanelStates[key]);
-				}
-			}
-		}
-		return this;
-	};
-
-	this.state._apply = function(newState) {
-		if (!ui._initialised) return msg('ERROR: UI not initialised');
-		if (typeof newState !== 'object') return msg('ERROR: invalid newState')
-		// this = ui.state
-		this.visible = newState.visible;
-		var newVisibility = newState.visibility;
-		if (typeof newState.viewCamera !== 'undefined') this.viewCamera = newState.viewCamera;
-
-		// take values from supplied visibility, but default to current
-		var k;
-		for (k in this.visibility) {
-			if (typeof newVisibility[k] !== 'undefined') this.visibility[k] = newVisibility[k];
-		}
-		k=null;
 
-		ui.applyVisibility(false);
 
-		if (typeof newState.panelStates === 'undefined') return true;	// nothing else left to do
+VizorUI.prototype.setupStateStoreEventListeners = function() {
+	var that = this;
+	var dom = this.dom;		// normally E2.dom
+	var state = this.state;
+	var visibility = state.visibility;
+	var $assets = dom.assetsLib, $presets = dom.presetsLib, $chat = dom.chatWindow;
+	var $patch_editor = dom.canvas_parent;
 
-		var ps = newState.panelStates;
-		if (typeof ps.chat !== 'undefined') {
-			var $chatPanel = ui.dom.chatWindow;
-			var chatState = ps.chat;
-			VizorUI.applyPanelState($chatPanel, chatState, ui.onChatToggleClicked.bind(ui));
-			var activeTabIsChat = (chatState.selectedTab === '#chatTab') || (chatState.selectedTab === 'chat');
-			if (activeTabIsChat && (chatState.h > 0)) {	// only apply height if the state had chat selected
-				ui.dom.chatWindow.height(chatState.h);
-				ui.onChatResize();
+	state
+		.on('changed:mode', function(mode) {
+			var inBuildMode = mode === uiMode.build
+			var inProgramMode = !inBuildMode
+			dom.btnBuildMode
+				.toggleClass('ui_on', inBuildMode)
+				.toggleClass('ui_off', inProgramMode);
+			dom.btnProgramMode
+				.toggleClass('ui_on', inProgramMode)
+				.toggleClass('ui_off', inBuildMode);
+
+			if (inProgramMode) {
+				dom.tabObjects.addClass('inactive ui_off');
+				dom.tabPresets.removeClass('inactive ui_off');
+			} else {
+				dom.tabObjects.removeClass('inactive ui_off');
+				dom.tabPresets.addClass('inactive ui_off');
 			}
-		}
-		if (typeof ps.presets !== 'undefined') {
-			VizorUI.applyPanelState(ui.dom.presetsLib, ps.presets, ui.onPresetsToggleClicked.bind(ui));
-		}
-		if (typeof ps.assets !== 'undefined') {
-			VizorUI.applyPanelState(ui.dom.assetsLib, ps.assets, ui.onAssetsToggleClicked.bind(ui));
-		}
 
-		ui.setWorldEditorMode(this.viewCamera === uiViewCam.world_editor);
+			dom.btnSavePatch.attr('disabled', inBuildMode);
+			dom.btnInspector.attr('disabled', inBuildMode);
+			dom.btnMove.attr('disabled',inProgramMode);
+			dom.btnScale.attr('disabled',inProgramMode);
+			dom.btnRotate.attr('disabled',inProgramMode);
+		})
+		.emit('changed:mode', state.mode);
```

Comment:
I couldn't think of any clearer way to do this - the handlers are set up and they are immediately initialised, with a few exceptions (important detail) - this process happens only once, otherwise I need to subscribe each handler to one more event each, or have the state emit every possible event, of which there are many, _or_ have the UI take in a copy of the state when it's first rendered, which I thought is overkill for this PR. I suggest a removeAllListeners at the start of this method, but I have no better or more elegant suggestions :/